### DISCUSSION TITLE
Order of externals

### DISCUSSION THREAD

**Discussion by arrrgi** [twpayne/chezmoi#3198] [Q&A] [answered]
Sorry to resurrect this old discussion [question](https://github.com/twpayne/chezmoi/discussions/1568). I couldn't find any reference to the order in which the .chezmoiexternal fits into the source state processing.

Is the the correct order?

1. run_before scripts in alphabetical/lexical order
2. .chezmoiexternals processed
3. Target states applied, files decrypted
4. run_after scripts in alphabetical/lexical order

I'm curious on whether I need to install "age" first with a script before Chezmoi applies any target states, or whether this can just be pulled as an external file reference 1st which Chezmoi can then use to decrypt any encrypted target state files as they are applied

**Comment by twpayne** [comment]
`.chezmoiexternals` are part of the source state and are read when the source state is read, i.e. before any scripts are run.

Note that chezmoi [has age encryption builtin](https://www.chezmoi.io/user-guide/encryption/age/#builtin-age-encryption) so you do not need to install age yourself. If you do have an `age` binary in your path then chezmoi will use that in preference to chezmoi's builtin age.

**Reply by arrrgi** [reply]
OK, so refining the order above:

1. .chezmoiexternals processed
2. run_before scripts
3. Target states applied, files decrypted
4. run_after scripts

With reference to **age** being built-in, I often see examples (including in the [documentation](https://www.chezmoi.io/user-guide/frequently-asked-questions/encryption/#how-do-i-configure-chezmoi-to-encrypt-files-but-only-request-a-passphrase-the-first-time-chezmoi-init-is-run)), which prompts the user for the passphrase, can this be substituted with **chezmoi** somehow?

```bash
#!/bin/sh

if [ ! -f "${HOME}/key.txt" ]; then
    age --decrypt --output "${HOME}/key.txt" "{{ .chezmoi.sourceDir }}/key.txt.age"
    chmod 600 "${HOME}/key.txt"
fi
```

becomes something like ...

```bash
#!/bin/sh

if [ ! -f "${HOME}/key.txt" ]; then
    chezmoi --decrypt --output "${HOME}/key.txt" "{{ .chezmoi.sourceDir }}/key.txt.age"
    chmod 600 "${HOME}/key.txt"
fi
```

Though this seems like it's not exactly safe to run chezmoi from chezmoi.

_Edit: I'm sure this is worth a question of it's own!_

**Reply by twpayne** [reply]
> With reference to **age** being built-in, I often see examples (including in the [documentation](https://www.chezmoi.io/user-guide/frequently-asked-questions/encryption/#how-do-i-configure-chezmoi-to-encrypt-files-but-only-request-a-passphrase-the-first-time-chezmoi-init-is-run)), which prompts the user for the passphrase, can this be substituted with **chezmoi** somehow?

Not yet.

chezmoi already has a [`decrypt` command](https://www.chezmoi.io/reference/commands/decrypt/). However, the builtin age command does not support passphrases. The use case you give is a reasonable justification for adding age passphrase support to chezmoi. I'll look into how feasible it is to add this.

Running chezmoi from chezmoi is generally OK as long as the nested chezmoi does not modify the source or target states. Specifically, `chezmoi decrypt` should be safe to run from chezmoi.

**Reply by arrrgi** [reply]
That's awesome - looking forward to seeing that included in the future!! 😎

For now I will fetch **age** with .chezmoiexternal first now that I'm clear on the order. Thanks!

**Reply by halostatue** [reply, marked as answer]
> OK, so refining the order above:
> 
> 1. .chezmoiexternals processed
> 2. run_before scripts
> 3. Target states applied, files decrypted
> 4. run_after scripts

I’m not quite sure that this is right, either. `.chezmoiexternals.$FORMAT` (the file) is processed with the rest of the state to prepare for execution. But chezmoi externals (the concept) are part of the applied target state.

I’m looking at the code and comparing it against [Application order](https://www.chezmoi.io/reference/application-order/).

1. (Not in current application order): check for config file freshness against the template
2. Read the source state.
   1. Read the source tree
   2. Read the externals
      - For archives, files, and archive files, this involves *optionally* refreshing the *cached result* or pulling the data from the external URL into the cache, and reading the state information.
      - Git externals are not refreshed at this point, nor are their states ever read (git externals delegate most state management tasks to git)
   3. Remove ignored targets from source state
   4. Generate removes based on `.chezmoiremove`
   5. Generate source state commands for git-repo externals (`git clone` or `git pull`).
3. Read the destination state.
4. Compute the target state.
5. Run run_before_ scripts in alphabetical order.
6. Update entries in the target state (files, directories, scripts, symlinks, etc.) in alphabetical order of their target name.
7. Run run_after_ scripts in alphabetical order.

I can’t be entirely certain, but it looks like externals are *applied* in the update phase (6) using the same target filename sort and *not* before `run_before_` scripts. They are not applied before `run_before_` scripts.

**Reply by arrrgi** [reply]
OK, so it would seem there's almost a circular dependency for unencrypting files with a passphrase. ie. you need to have age already installed using `run_before_` scripts.

I've unmarked the answer for now while it's not completely clear the best approach to using externals vs scripts to get **age** available for decrypting files with a passphrase.

Let's see what @twpayne clarifies.

**Reply by halostatue** [reply]
I think that when #3199 is merged and released, this *particular* issue will be moot, but I’ve opened #3201 to try to clarify what I’ve found.

I think reentrancy detection will be somewhat hard to do, but if we do that, I’d go strict at first (permitting only known safe items that do not read or write to the state files). *Some* things which would nominally be non-reentrant could *probably* be made safely reentrant (I think one should be able to call `chezmoi execute-template` from a script being run by `chezmoi apply`, but *why* you would do so is an entirely different matter).

**Reply by arrrgi** [reply]
Thanks for the doc update @halostatue
